ykrt/src/compile/jitc_yk/trace_builder.rs

```diff
@@ -301,42 +314,76 @@ impl TraceBuilder {
                     dyn_elem_sizes,
                     ..
                 } => {
-                    self.handle_ptradd(bid, iidx, ptr, *const_off, dyn_elem_counts, dyn_elem_sizes)
+                    self.handle_ptradd(
+                        bid,
+                        iidx,
+                        ptr,
+                        *const_off,
+                        dyn_elem_counts,
+                        dyn_elem_sizes,
+                    )?;
+                    Ok(None)
                 }
                 aot_ir::Inst::BinaryOp { lhs, binop, rhs } => {
-                    self.handle_binop(bid, iidx, *binop, lhs, rhs)
+                    self.handle_binop(bid, iidx, *binop, lhs, rhs)?;
+                    Ok(None)
                 }
                 aot_ir::Inst::ICmp { lhs, pred, rhs, .. } => {
-                    self.handle_icmp(bid, iidx, lhs, pred, rhs)
+                    self.handle_icmp(bid, iidx, lhs, pred, rhs)?;
+                    Ok(None)
                 }
                 aot_ir::Inst::CondBr {
                     cond,
                     true_bb,
                     safepoint,
                     ..
-                } => self.handle_condbr(bid, safepoint, nextbb.as_ref().unwrap(), cond, true_bb),
+                } => {
+                    self.handle_condbr(bid, safepoint, nextbb.as_ref().unwrap(), cond, true_bb)?;
+                    Ok(None)
+                }
                 aot_ir::Inst::Cast {
                     cast_kind,
                     val,
                     dest_tyidx,
-                } => self.handle_cast(bid, iidx, cast_kind, val, dest_tyidx),
-                aot_ir::Inst::Ret { val } => self.handle_ret(bid, iidx, val),
+                } => {
+                    self.handle_cast(bid, iidx, cast_kind, val, dest_tyidx)?;
+                    Ok(None)
+                }
+                aot_ir::Inst::Ret { val } => {
+                    let retframe = self.handle_ret(bid, iidx, val)?;
+                    // In the case of a `ret` terminator, we need to communicate back to the caller
+                    // the block that should become the new `prevbid` going forward.
+                    //
+                    // This is nasty.
+                    let ret_prevbb: Option<Option<BBlockId>> =
+                        if let Some(ref ci) = retframe.callinst {
+                            Some(Some(BBlockId::new(ci.funcidx(), ci.bbidx())))
+                        } else {
+                            Some(None)
+                        };
+                    // Early return OK because `ret` is a terminator and there can be no further
+                    // instructions in the block.
+                    return Ok(ret_prevbb);
```

Comment:
AFAICS this is the only thing we need to add: the `return` short-circuits the implict `Ok(None)` at the end of the function?